Classes/Middleware/FrontendUserAuthenticator.php

```diff
@@ -0,0 +1,158 @@
+ * All rights reserved
+ *
+ * This script is part of the TYPO3 project. The TYPO3 project is
+ * free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The GNU General Public License can be found at
+ * http://www.gnu.org/copyleft/gpl.html.
+ * A copy is found in the text file GPL.txt and important notices to the license
+ * from the author is found in LICENSE.txt distributed with these scripts.
+ *
+ *
+ * This script is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * This copyright notice MUST APPEAR in all copies of the script!
+ ***************************************************************/
+
+use ApacheSolrForTypo3\Solr\Access\Rootline;
+use ApacheSolrForTypo3\Solr\IndexQueue\FrontendHelper\AuthorizationService;
+use ApacheSolrForTypo3\Solr\IndexQueue\PageIndexerRequest;
+use ApacheSolrForTypo3\Solr\IndexQueue\PageIndexerRequestHandler;
+use ApacheSolrForTypo3\Solr\System\Logging\SolrLogManager;
+use ApacheSolrForTypo3\Solr\Util;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use TYPO3\CMS\Core\Context\Context;
+use TYPO3\CMS\Core\Context\UserAspect;
+use TYPO3\CMS\Core\Http\JsonResponse;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
+use TYPO3\CMS\Frontend\Authentication\FrontendUserAuthentication;
+
+/**
+ * Class FrontendUserAuthenticator is responsible to fake a frontend user and fe_groups.
+ * It makes possible to Index access restricted pages and content elements.
+ */
+class FrontendUserAuthenticator implements MiddlewareInterface
+{
+
+    /**
+     * @var Context
+     */
+    protected $context;
+
+    /**
+     * FrontendUserAuthorizator constructor.
+     *
+     * @param Context|null $context
+     * @noinspection PhpUnused
+     */
+    public function __construct(?Context $context = null)
+    {
+        $this->context = $context ?? GeneralUtility::makeInstance(Context::class);
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @param RequestHandlerInterface $handler
+     * @return ResponseInterface
+     * @noinspection PhpUnused
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        if (Util::getIsTYPO3VersionBelow10()
+            || !$request->hasHeader(PageIndexerRequest::SOLR_INDEX_HEADER)
+        ) {
+            return $handler->handle($request);
+        }
+
+        // disable TSFE cache for TYPO3 v10
+        $request = $request->withAttribute('noCache', true);
```

Comment:
this won't work if no_cache is disabled via `LocalConfiguration.php`
I propose disabling the cache with a middleware/hook when indexing, something like https://github.com/studiomitte/powermail_freebie/blob/master/Classes/Hooks/CacheHook.php

Reply:
I'll create a new issue for that. We should clean up the old stuff anyway, there are multiple things, which try to disable caches.